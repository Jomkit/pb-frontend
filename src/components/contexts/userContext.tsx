import { createContext } from "react";
import { IUser } from "../../types";

const userContext = createContext({} as IUser);

export default userContext;