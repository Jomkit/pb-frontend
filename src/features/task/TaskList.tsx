import { IProject, ITask } from '../../types'
import Task from './Task'
import AddBtn from '../../components/AddBtn'
import { useState } from 'react';
import NewTaskForm from './NewTaskForm';
import Popup from '../../components/Popup';
import Card from '../../components/Card';

const TaskList = ({loading, projects, tasks}: {loading: boolean, projects: IProject[], tasks: ITask[]}) => {
  const [popupOpen, setPopupOpen] = useState(false);
  
  const addTask = () => {
    console.log("Add task");
    setPopupOpen(true);
  }
  
  const close = () => {
    setPopupOpen(false);
  }
  
  return loading ? <p>Loading...</p> : (
    <div className='w-3/4 m-auto'>
        <AddBtn handleClick={addTask} btnName="Task"/>
        {popupOpen && (
          <Popup close={close}>
            <NewTaskForm projects={projects} closePopup={close} />
          </Popup>
        )}
        {
          tasks.length === 0 ? 
          <Card>
            <h2>No Tasks</h2>
            <p>Get started by creating a new task.</p>
          </Card>
          :
          tasks.map((task: ITask) => (
            <Task key={task.id} task={task} projectName={projects.find((project: IProject) => project.id === task.projectId)!.name}/>
          ))
        }
    </div>
  )
}

export default TaskList