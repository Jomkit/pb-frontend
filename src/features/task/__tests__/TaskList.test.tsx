import { render } from "@testing-library/react"
import TaskList from "../TaskList"
import userContext from "../../../components/contexts/userContext";
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";

const testTasks = [
    {
        id: "1",
        name: "testTask 1",
        projectId: "1"
    },
    {
        id: "2",
        name: "testTask 2",
        projectId: "1"
    },
    {
        id: "3",
        name: "testTask 3",
        projectId: "1"
    }
]

const testUser = {
    id: 1,
    username: "testUser",
    firstName: "test",
    lastName: "test",
    email: "test@email.com"
}

let loading = true;

afterEach(() => {
    loading = true;
})

it("should render without crashing", () => {
    render(<TaskList loading={loading} tasks={testTasks} />);
})

it("should match snapshot of loading view", () => {
    const { asFragment } = render(<TaskList loading={loading} tasks={testTasks} />);
    expect(asFragment()).toMatchSnapshot();
})

it("should match snapshot of loading complete", async () => {
    loading = false;
    const { asFragment, findAllByText } = render(
        <MemoryRouter>
            <userContext.Provider value={testUser}>
            <TaskList loading={loading} tasks={testTasks} />
            </userContext.Provider>

        </MemoryRouter>
    );

    const allTestTasks = await findAllByText("testTask", { exact: false });
    expect(allTestTasks.length).toBe(3);
    expect(asFragment()).toMatchSnapshot();
})

it("should bring up popup on clicking New Task button", async () => {
    loading = false;
    const user = userEvent.setup();
    const { findByText } = render(
        <MemoryRouter>
            <userContext.Provider value={testUser}>
            <TaskList loading={loading} tasks={testTasks} />
            </userContext.Provider>
        </MemoryRouter>
    );
        
    const newTaskBtn = await findByText("New Task");
    expect(newTaskBtn).toBeInTheDocument();

    await user.click(newTaskBtn);
    const formTitle = await findByText("Create New Task");
    expect(formTitle).toBeInTheDocument();

    const closeButton = await findByText("✕");
    expect(closeButton).toBeInTheDocument();
})

it("should close popup on clicking close button", async () => {
    loading = false;
    const user = userEvent.setup();
    const { findByText } = render(
        <MemoryRouter>
            <userContext.Provider value={testUser}>
            <TaskList loading={loading} tasks={testTasks} />
            </userContext.Provider>
        </MemoryRouter>
    );
        
    const newTaskBtn = await findByText("New Task");
    expect(newTaskBtn).toBeInTheDocument();

    await user.click(newTaskBtn);
    const formTitle = await findByText("Create New Task");
    expect(formTitle).toBeInTheDocument();

    const closeButton = await findByText("✕");
    expect(closeButton).toBeInTheDocument();

    await user.click(closeButton);
    expect(formTitle).not.toBeInTheDocument();
})