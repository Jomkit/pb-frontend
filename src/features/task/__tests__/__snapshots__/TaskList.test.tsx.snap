// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot of loading complete 1`] = `
<DocumentFragment>
  <div
    class="w-3/4 m-auto"
  >
    <div
      class="text-start"
    >
      <button
        class="btn bg-green-400 border-solid border-2 border-black"
      >
        New Task
      </button>
    </div>
    <div
      class="w-full mx-auto mb-2 p-6 bg-white border-solid border-2 border-black rounded-lg divide-y"
    >
      <div
        class="text-center"
      >
        <h2>
          testTask 1
        </h2>
        <h4>
          Project: testProject 1
        </h4>
      </div>
      <a
        href="/tasks/1"
      >
        <p
          class="btn bg-blue-500 text-white"
        >
          Click for more info
        </p>
      </a>
      <div
        class="m-2"
      >
        <div
          class="w-full mx-auto mb-2 p-6 bg-white border-solid border-2 border-black rounded-lg divide-y"
        >
          <div
            class="text-center"
          >
            <h2 />
            <h4 />
          </div>
          <h2
            class="text-start"
          >
            Timer
          </h2>
          <div
            class="flex justify-between"
          >
            <div
              class="text-center"
            >
              <h3
                class="text-2xl bg-slate-200 w-fit m-auto px-4 py-1 rounded border-2 border-slate-500"
              >
                25:00
              </h3>
            </div>
            <div
              id="set-timer"
            >
              <button
                class="btn bg-green-400"
                id="start-btn"
              >
                <i
                  class="fa-solid fa-play"
                />
              </button>
              <button
                class="btn bg-red-500"
                id="restart-btn"
              >
                <i
                  class="fa-solid fa-rotate-right"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="w-full mx-auto mb-2 p-6 bg-white border-solid border-2 border-black rounded-lg divide-y"
    >
      <div
        class="text-center"
      >
        <h2>
          testTask 2
        </h2>
        <h4>
          Project: testProject 1
        </h4>
      </div>
      <a
        href="/tasks/2"
      >
        <p
          class="btn bg-blue-500 text-white"
        >
          Click for more info
        </p>
      </a>
      <div
        class="m-2"
      >
        <div
          class="w-full mx-auto mb-2 p-6 bg-white border-solid border-2 border-black rounded-lg divide-y"
        >
          <div
            class="text-center"
          >
            <h2 />
            <h4 />
          </div>
          <h2
            class="text-start"
          >
            Timer
          </h2>
          <div
            class="flex justify-between"
          >
            <div
              class="text-center"
            >
              <h3
                class="text-2xl bg-slate-200 w-fit m-auto px-4 py-1 rounded border-2 border-slate-500"
              >
                25:00
              </h3>
            </div>
            <div
              id="set-timer"
            >
              <button
                class="btn bg-green-400"
                id="start-btn"
              >
                <i
                  class="fa-solid fa-play"
                />
              </button>
              <button
                class="btn bg-red-500"
                id="restart-btn"
              >
                <i
                  class="fa-solid fa-rotate-right"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="w-full mx-auto mb-2 p-6 bg-white border-solid border-2 border-black rounded-lg divide-y"
    >
      <div
        class="text-center"
      >
        <h2>
          testTask 3
        </h2>
        <h4>
          Project: testProject 1
        </h4>
      </div>
      <a
        href="/tasks/3"
      >
        <p
          class="btn bg-blue-500 text-white"
        >
          Click for more info
        </p>
      </a>
      <div
        class="m-2"
      >
        <div
          class="w-full mx-auto mb-2 p-6 bg-white border-solid border-2 border-black rounded-lg divide-y"
        >
          <div
            class="text-center"
          >
            <h2 />
            <h4 />
          </div>
          <h2
            class="text-start"
          >
            Timer
          </h2>
          <div
            class="flex justify-between"
          >
            <div
              class="text-center"
            >
              <h3
                class="text-2xl bg-slate-200 w-fit m-auto px-4 py-1 rounded border-2 border-slate-500"
              >
                25:00
              </h3>
            </div>
            <div
              id="set-timer"
            >
              <button
                class="btn bg-green-400"
                id="start-btn"
              >
                <i
                  class="fa-solid fa-play"
                />
              </button>
              <button
                class="btn bg-red-500"
                id="restart-btn"
              >
                <i
                  class="fa-solid fa-rotate-right"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should match snapshot of loading view 1`] = `
<DocumentFragment>
  <p>
    Loading...
  </p>
</DocumentFragment>
`;
